### PR TITLE
test: ionViewWillEnter method in add-edit-per-diem part 4

### PR DESCRIPTION
**Pull request by suyashpatil78**
### <samp>🤖 Generated by Copilot at 2be94ee</samp>

This pull request improves the testing and validation of the per diem form feature. It adds and updates mock data for per diem form values and reports. It also refactors the payment mode validation logic and updates the per diem form value interface.

### <samp>🤖 Generated by Copilot at 2be94ee</samp>

> _Sing, O Muse, of the skilled coder who wrought_
> _Many a mock data for the per diem form,_
> _And tested well the logic and the plot_
> _Of adding and editing expenses in the storm._

### <samp>🤖 Generated by Copilot at 2be94ee</samp>

*  Add mock data for per diem form values and a draft report with per diem expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-ca7b146ba7bb148d1688424c113ba77ac0191512d1b73bc1193ec18191761cbfR1-R66), [link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-d67cba8ad3a3712af6f510fa26bd7076e3e8bcd472aef15dcd35f0a4cdc32eb2R1534-R1610))
* Update the types of `from_dt` and `to_dt` fields in `PerDiemFormValue` interface to match the mock data and the form values ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-37857451b82e7e2d0d887408f452de68759a80e2160b3366e5c23b10c2efcbb0L23-R24))
* Add `flush` function to the import from `@angular/core/testing` to flush the microtasks queue in the fakeAsync tests ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dL1-R1))
* Add mock data for an expense without a cost center, different types of accounts, multiple advance accounts, and an unflattened account to the import from `src/app/core/mock-data/unflattened-txn.data` and `src/app/core/test-data/accounts.service.spec.data` to test the autofill and validity of cost center and payment mode fields in the per diem form ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dL38-R38), [link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dL53-R62))
* Import mock data for per diem form values from `src/app/core/mock-data/per-diem-form-value.data.ts` to test the autofill of different fields in the per diem form ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dR81-R89))
* Add test cases for checking the balance availability in the advance accounts, the autofill of the per diem form, and the payment mode validity using the mock data and the services ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dR739-R872))
* Add a method `isPaymentModeValid` in class `AddEditPerDiemPage` in `src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts` that returns an observable of boolean value indicating whether the payment mode is valid or not ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dR802-R823))
* Replace the previous logic of `paymentModeInvalid$` with the call to the `isPaymentModeValid` method in class `AddEditPerDiemPage` in `src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts` to refactor the code and avoid duplication of logic ([link](https://github.com/fylein/fyle-mobile-app/pull/2294/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL1549-R1571))

## Clickup
https://app.clickup.com/t/85ztk93eg

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/core/models/per-diem-form-values.model.ts
Comment (Chethan-Fyle) [inline src/app/core/models/per-diem-form-values.model.ts:1]: File name is not matching with interface name (`per-diem-form-values` and `PerDiemFormValue`).

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts [hunk 1]
```diff
@@ -724,6 +736,140 @@ export function TestCases2(getTestBed) {
...
+
+      it('should set isBalanceAvailableInAnyAdvanceAccount$ correctly if paymentMode changed and paymentMode.acc.type equals PERSONAL account', (done) => {
+        accountsService.getEtxnSelectedPaymentMode.and.returnValue(accountsData[0]);
+        accountsService.getEMyAccounts.and.returnValue(of(accountsData));
+        component.ionViewWillEnter();
+        component.isBalanceAvailableInAnyAdvanceAccount$.subscribe((res) => {
+          expect(accountsService.getEMyAccounts).toHaveBeenCalledTimes(1);
+          expect(res).toBeTrue();
+        });
+        done();
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts:748]: Shouldn't this be inside the `subscribe()` callback method?

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 2]
```diff
@@ -799,6 +799,28 @@ export class AddEditPerDiemPage implements OnInit {
...
   }
 
+  isPaymentModeValid(): Observable<boolean> {
+    return iif(() => !!(this.activatedRoute.snapshot.params.id as string), this.etxn$, of(null)).pipe(
+      map((etxn) => {
+        const formValue = this.getFormValues();
+        if (formValue.paymentMode.acc.type === AccountType.ADVANCE) {
+          if (
+            etxn &&
+            etxn.tx.id &&
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:809]: `if (etxn?.tx.id && ...)`

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 3]
```diff
@@ -799,6 +799,28 @@ export class AddEditPerDiemPage implements OnInit {
...
+            formValue.paymentMode?.acc?.id === etxn.tx.source_account_id &&
+            etxn.tx.state !== 'DRAFT'
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:811]: Create enum for transaction state.

**Review by Chethan-Fyle — APPROVED** [review]
Minor Comments.

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts (hunk 1 above)
Comment (OmkarJ13) [reply src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts:748]: +

**Review by OmkarJ13 — APPROVED** [review]

File: src/app/core/models/per-diem-form-values.model.ts
Comment (suyashpatil78) [reply src/app/core/models/per-diem-form-values.model.ts:1]: changing this!

**Pull request merged**